src/expr/node_traversal.cpp

```diff
@@ -0,0 +1,182 @@
+/*********************                                                        */
+/*! \file node_traversal.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Alex Ozdemir
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2019 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Iterators for traversing nodes.
+ **/
+
+#include "node_traversal.h"
+
+#include <iostream>
+
+namespace CVC4 {
+
+NodeDfsIterator::NodeDfsIterator(TNode n, bool postorder)
+    : d_stack{n}, d_visited(), d_postorder(postorder)
+{
+}
+
+NodeDfsIterator::NodeDfsIterator(bool postorder)
+    : d_stack(), d_visited(), d_postorder(postorder)
+{
+}
+
+NodeDfsIterator& NodeDfsIterator::operator++()
+{
+  std::cout << "post-increment: " << d_stack << std::endl;
+  // If we were just constructed, advance to first visit
+  advanceUntilVisitIfJustConstructed();
+
+  // Advance past our current visit
+  if (d_postorder)
+  {
+    finishPostVisit();
+  }
+  else
+  {
+    finishPreVisit();
+  }
+
+  // Make sure we're at an appropriate visit
+  advanceUntilVisit();
+  return *this;
+}
+
+NodeDfsIterator NodeDfsIterator::operator++(int)
+{
+  NodeDfsIterator copyOfOld(*this);
+  ++*this;
+  return copyOfOld;
+}
+
+TNode& NodeDfsIterator::operator*()
+{
+  std::cout << "deref         : " << d_stack << std::endl;
+  // If we were just constructed, advance to first visit
+  advanceUntilVisitIfJustConstructed();
+
+  return d_stack.back();
+}
+
+bool NodeDfsIterator::operator==(const NodeDfsIterator& other) const
+{
+  // The stack uniquely represents traversal state. We need not use the
+  // scheduled node set. We also ignore the order: users should not compare
+  // nodes of different order.
+  return d_stack == other.d_stack;
+}
+
+bool NodeDfsIterator::operator!=(const NodeDfsIterator& other) const
+{
+  return !(*this == other);
+}
+
+void NodeDfsIterator::advance()
+{
+  Assert(!d_stack.empty());
+  TNode back = d_stack.back();
+  auto visitEntry = d_visited.find(back);
+  if (visitEntry == d_visited.end())
+  {
+    // if we haven't pre-visited this node, pre-visit it
+    finishPreVisit();
+  }
+  else if (visitEntry->second)
+  {
+    // if we've already post-visited this node: skip it
+    d_stack.pop_back();
+  }
+  else
+  {
+    // otherwise, post-visit it
+    finishPostVisit();
+  }
+}
+
+void NodeDfsIterator::advanceUntilVisit()
+{
+  // While a node is enqueued ..
+  while (d_postorder ? !atPostVisit() : !atPreVisit())
+  {
+    advance();
```

Comment:
Since `advance()` is only used here, I would inline the code. It makes the traversal less confusing. You should be also able to inline `atPostVisit()` and `atPreVisit()` then you don't have to find `visitEntry` multiple times and you can directly break out of the loop.

Reply:
Done in a049c682f